File: crawler/security/http_executor.py
"""Isolated secure HTTP executor for TASK-022D-3.

The executor combines pinned addressing, policy, redirect planning, runtime
budgets, and the shared concurrency limiter. It does not use requests/httpx
and never follows redirects automatically.
"""

from __future__ import annotations

import time
from typing import Any, Callable

from crawler.security.bounded_io import BoundedIOError
from crawler.security.concurrency_limiter import ConcurrencyError, ConcurrencyLimiter
from crawler.security.http_transport import (
    SecureHTTPError,
    SecureHTTPRequest,
    SecureHTTPResponse,
    build_request_bytes,
    read_raw_response,
    validate_request,
)
from crawler.security.outbound_policy import decide
from crawler.security.pinned_connection import PinnedConnectionError, build_pinned_target, connect_pinned
from crawler.security.redirect_policy import RedirectPolicyError, plan_redirect_hop
from crawler.security.tls_policy import TLSPolicyError, secure_ssl_context, wrap_client_socket
from crawler.security.transport_budget import BUDGET, BudgetError, budget_for, remaining_deadline

_REDIRECT_STATUSES = frozenset({301, 302, 303, 307, 308})
_SENSITIVE_HEADERS = frozenset({"authorization", "proxy-authorization", "cookie"})


class SecureHTTPExecutor:
    def __init__(
        self,
        policy,
        resolver,
        limiter: ConcurrencyLimiter,
        profile: str,
        *,
        socket_factory: Callable[[int, int], Any] | None = None,
    ):
        if limiter is None:
            raise ValueError("limiter is required")
        self.policy = policy
        self.resolver = resolver
        self.limiter = limiter
        self.profile = profile
        self.socket_factory = socket_factory
        budget_for(profile)

    def execute(self, request: SecureHTTPRequest) -> SecureHTTPResponse:
        validate_request(request)
        start_ms = int(time.monotonic() * 1000)
        current_url = request.url
        current_normalized = request.url
        headers = request.headers
        body = request.body
        hops = 0
        previous_host: str | None = None

        for hop_count in range(1, BUDGET.max_redirects + 2):
            if hop_count == 1:
                decision = decide(current_url, self.policy, resolver=self.resolver)
                if not decision.allowed:
                    raise SecureHTTPError(decision.reason_code)
                current_normalized = decision.normalized.normalized_url
                target = build_pinned_target(decision, self.policy)
            else:
                try:
                    hop = plan_redirect_hop(
                        current_normalized,
                        current_url,
                        self.policy,
                        resolver=self.resolver,
                        hop_count=hop_count - 1,
                    )
                except RedirectPolicyError as exc:
                    raise SecureHTTPError(exc.reason_code) from exc
                if previous_host is not None and hop.pinned_target.normalized_host != previous_host:
                    headers = _strip_sensitive(headers)
                current_normalized = hop.url
                target = hop.pinned_target
                hops += 1

            current_host = target.normalized_host
            remaining = self._remaining(start_ms)
            try:
                lease = self.limiter.acquire(target.normalized_host)
            except ConcurrencyError as exc:
                raise SecureHTTPError(exc.reason_code) from exc
            conn = None
            try:
                connect_timeout = min(BUDGET.connect_timeout_ms, remaining)
                conn = connect_pinned(
                    target,
                    socket_factory=self.socket_factory,
                    timeout=connect_timeout / 1000.0,
                )
                if target.scheme == "https":
                    tls_remaining = self._remaining(start_ms)
                    tls_timeout = min(BUDGET.tls_timeout_ms, tls_remaining)
                    conn.settimeout(tls_timeout / 1000.0)
                    context = secure_ssl_context()
                    conn = wrap_client_socket(context, conn, target.server_name)
                request_bytes = build_request_bytes(
                    SecureHTTPRequest(
                        profile=self.profile,
                        method=request.method,
                        url=current_normalized,
                        headers=headers,
                        body=body,
                    ),
                    target.host_header,
                )
                write_remaining = self._remaining(start_ms)
                conn.settimeout(min(BUDGET.response_header_timeout_ms, write_remaining) / 1000.0)
                conn.sendall(request_bytes)
                read_remaining = self._remaining(start_ms)
                status, response_headers, response_body = read_raw_response(
                    conn,
                    request.method,
                    self.profile,
                    start_ms,
                    read_remaining,
                    _clock,
                )
                if status in _REDIRECT_STATUSES:
                    location = ""
                    for name, value in response_headers:
                        if name.lower() == "location":
                            location = value
                    if not location:
                        raise SecureHTTPError("redirect_location_missing")
                    if request.method == "POST":
                        raise SecureHTTPError("redirect_body_replay_not_allowed")
                    previous_host = current_host
                    current_url = location
                    continue
                return SecureHTTPResponse(
                    status_code=status,
                    final_url=current_normalized,
                    headers=response_headers,
                    body=response_body,
                    redirect_hops=hops,
                )
            except (SecureHTTPError, BoundedIOError, PinnedConnectionError, TLSPolicyError, BudgetError) as exc:
                raise SecureHTTPError(getattr(exc, "reason_code", "transport_failed")) from exc
            except TimeoutError as exc:
                raise SecureHTTPError("timeout") from exc
            except OSError as exc:
                raise SecureHTTPError("transport_failed") from exc
            finally:
                if conn is not None:
                    try:
                        conn.close()
                    except Exception:
                        pass
                lease.release()
        raise SecureHTTPError("redirect_limit_exceeded")

    def _remaining(self, start_ms: int) -> int:
        elapsed = int(time.monotonic() * 1000) - start_ms
        try:
            remaining = remaining_deadline(self.profile, elapsed)
        except BudgetError as exc:
            raise SecureHTTPError(exc.reason_code) from exc
        if remaining <= 0:
            raise SecureHTTPError("total_timeout_exceeded")
        return remaining


def _strip_sensitive(headers: tuple[tuple[str, str], ...]) -> tuple[tuple[str, str], ...]:
    return tuple(
        (name, value)
        for name, value in headers
        if name.lower() not in _SENSITIVE_HEADERS
    )


def _clock() -> int:
    return int(time.monotonic() * 1000)

File: go-spider/internal/security/http_executor.go
package security

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"net"
	"strings"
	"time"
)

// SecureHTTPExecutor combines pinned transport, redirect planning, budgets,
// and the shared concurrency limiter for isolated HTTP/1.1 execution.
type SecureHTTPExecutor struct {
	policy   OutboundPolicy
	resolver Resolver
	limiter  *ConcurrencyLimiter
	profile  BudgetProfile
	dial     DialContextFunc
	rootCAs  *x509.CertPool
}

// NewSecureHTTPExecutor constructs the isolated executor.
func NewSecureHTTPExecutor(policy OutboundPolicy, resolver Resolver, limiter *ConcurrencyLimiter, profile BudgetProfile, dial DialContextFunc, rootCAs *x509.CertPool) *SecureHTTPExecutor {
	return &SecureHTTPExecutor{policy: policy, resolver: resolver, limiter: limiter, profile: profile, dial: dial, rootCAs: rootCAs}
}

func (e *SecureHTTPExecutor) Execute(req SecureHTTPRequest) (SecureHTTPResponse, error) {
	if e.limiter == nil {
		return SecureHTTPResponse{}, secureHTTPFail("limiter_required")
	}
	if err := ValidateSecureRequest(req); err != nil {
		return SecureHTTPResponse{}, err
	}
	start := time.Now()
	currentURL := req.URL
	currentNormalized := req.URL
	headers := req.Headers
	hops := 0
	previousHost := ""

	for hopCount := 1; hopCount <= NewTransportBudget().MaxRedirects()+1; hopCount++ {
		var target PinnedTarget
		if hopCount == 1 {
			decision, err := DecidePolicy(currentURL, e.policy, e.resolver)
			if err != nil {
				return SecureHTTPResponse{}, secureReason(err)
			}
			if !decision.Allowed {
				return SecureHTTPResponse{}, secureHTTPFail(decision.ReasonCode)
			}
			currentNormalized = decision.Normalized.NormalizedURL
			target, err = NewPinnedTarget(decision, e.policy)
			if err != nil {
				return SecureHTTPResponse{}, secureReason(err)
			}
		} else {
			hop, err := PlanRedirectHop(currentNormalized, currentURL, e.policy, e.resolver, hopCount-1)
			if err != nil {
				return SecureHTTPResponse{}, secureReason(err)
			}
			if previousHost != "" && hop.Target.NormalizedHost() != previousHost {
				headers = stripSensitiveHeaders(headers)
			}
			currentNormalized = hop.URL
			target = hop.Target
			hops++
		}

		currentHost := target.NormalizedHost()
		elapsed := int64(time.Since(start).Milliseconds())
		remaining, err := RemainingDeadline(e.profile, int(elapsed))
		if err != nil {
			return SecureHTTPResponse{}, secureReason(err)
		}
		if remaining <= 0 {
			return SecureHTTPResponse{}, secureHTTPFail("total_timeout_exceeded")
		}
		lease, err := e.limiter.Acquire(target.NormalizedHost())
		if err != nil {
			return SecureHTTPResponse{}, secureReason(err)
		}
		conn, err := e.dialForTarget(target, remaining, start)
		if err != nil {
			lease.Release()
			return SecureHTTPResponse{}, err
		}
		status, responseHeaders, responseBody, rerr := e.roundTrip(conn, req, headers, target, currentNormalized, start, remaining)
		_ = conn.Close()
		lease.Release()
		if rerr != nil {
			return SecureHTTPResponse{}, rerr
		}
		if status == 301 || status == 302 || status == 303 || status == 307 || status == 308 {
			location := ""
			for _, header := range responseHeaders {
				if strings.EqualFold(header.Name, "Location") {
					location = header.Value
				}
			}
			if location == "" {
				return SecureHTTPResponse{}, secureHTTPFail("redirect_location_missing")
			}
			if req.Method == "POST" {
				return SecureHTTPResponse{}, secureHTTPFail("redirect_body_replay_not_allowed")
			}
			previousHost = currentHost
			currentURL = location
			continue
		}
		return SecureHTTPResponse{StatusCode: status, FinalURL: currentNormalized, Headers: responseHeaders, Body: responseBody, RedirectHops: hops}, nil
	}
	return SecureHTTPResponse{}, secureHTTPFail("redirect_limit_exceeded")
}

func (e *SecureHTTPExecutor) dialForTarget(target PinnedTarget, remaining int, start time.Time) (net.Conn, error) {
	if remaining <= 0 {
		return nil, secureHTTPFail("total_timeout_exceeded")
	}
	budget := NewTransportBudget()
	connectTimeout := budget.ConnectTimeoutMS()
	if remaining < connectTimeout {
		connectTimeout = remaining
	}
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(connectTimeout)*time.Millisecond)
	defer cancel()
	conn, err := DialPinned(ctx, target, e.dial)
	if err != nil {
		return nil, secureReason(err)
	}
	if target.Scheme() != "https" {
		return conn, nil
	}
	elapsed := int64(time.Since(start).Milliseconds())
	remainingNow, err := RemainingDeadline(e.profile, int(elapsed))
	if err != nil {
		_ = conn.Close()
		return nil, secureReason(err)
	}
	if remainingNow <= 0 {
		_ = conn.Close()
		return nil, secureHTTPFail("total_timeout_exceeded")
	}
	tlsTimeout := budget.TLSTimeoutMS()
	if remainingNow < tlsTimeout {
		tlsTimeout = remainingNow
	}
	tlsConfig := &tls.Config{
		ServerName:         target.ServerName(),
		InsecureSkipVerify: false,
	}
	if e.rootCAs != nil {
		tlsConfig.RootCAs = e.rootCAs.Clone()
	}
	tlsConn := tls.Client(conn, tlsConfig)
	_ = tlsConn.SetDeadline(time.Now().Add(time.Duration(tlsTimeout) * time.Millisecond))
	if err := tlsConn.Handshake(); err != nil {
		_ = conn.Close()
		if ne, ok := err.(net.Error); ok && ne.Timeout() {
			return nil, secureHTTPFail("tls_timeout")
		}
		return nil, secureHTTPFail("transport_failed")
	}
	_ = tlsConn.SetDeadline(time.Time{})
	return tlsConn, nil
}

func (e *SecureHTTPExecutor) roundTrip(conn net.Conn, req SecureHTTPRequest, headers []Header, target PinnedTarget, normalizedURL string, start time.Time, remaining int) (int, []Header, []byte, error) {
	requestBytes, err := BuildRequestBytes(SecureHTTPRequest{Profile: req.Profile, Method: req.Method, URL: normalizedURL, Headers: headers, Body: req.Body}, target.HostHeader())
	if err != nil {
		return 0, nil, nil, err
	}
	_ = conn.SetWriteDeadline(time.Now().Add(time.Duration(remaining) * time.Millisecond))
	if _, err := conn.Write(requestBytes); err != nil {
		return 0, nil, nil, secureHTTPFail("transport_failed")
	}
	elapsedBeforeRead := int64(time.Since(start).Milliseconds())
	remainingNow, err := RemainingDeadline(e.profile, int(elapsedBeforeRead))
	if err != nil {
		return 0, nil, nil, secureReason(err)
	}
	if remainingNow <= 0 {
		return 0, nil, nil, secureHTTPFail("total_timeout_exceeded")
	}
	clock := func() int64 { return int64(time.Since(start).Milliseconds()) }
	return ReadRawResponse(conn, req.Method, e.profile, elapsedBeforeRead, int64(remainingNow), clock)
}

func stripSensitiveHeaders(headers []Header) []Header {
	out := make([]Header, 0, len(headers))
	for _, header := range headers {
		switch strings.ToLower(header.Name) {
		case "authorization", "proxy-authorization", "cookie":
			continue
		}
		out = append(out, header)
	}
	return out
}

func secureReason(err error) error {
	if err == nil {
		return nil
	}
	if _, ok := err.(*SecureHTTPError); ok {
		return err
	}
	return secureHTTPFail(reasonCodeFromError(err))
}

func reasonCodeFromError(err error) string {
	if e, ok := err.(*SecureHTTPError); ok {
		return e.ReasonCode
	}
	if e, ok := err.(*BoundedIOError); ok {
		return e.ReasonCode
	}
	if e, ok := err.(*BudgetError); ok {
		return e.ReasonCode
	}
	if e, ok := err.(*PinnedConnectionError); ok {
		return e.ReasonCode
	}
	if e, ok := err.(*RedirectPolicyError); ok {
		return e.ReasonCode
	}
	if e, ok := err.(*DNSValidationError); ok {
		return e.ReasonCode
	}
	if e, ok := err.(*URLNormalizationError); ok {
		return e.ReasonCode
	}
	if e, ok := err.(*ConcurrencyError); ok {
		return e.ReasonCode
	}
	return "transport_failed"
}
